### PR TITLE
new entityExtractor atom

### PR DESCRIPTION
**Pull request by bmichele**
Modify entityExtractor atomic so that accepts an argument entity_type with a single descriptor:

```
query: "helsinki"
entityExtractor(entityTypes="CITY_FI") returns score 1
entityExtractor(entityTypes="NAMES") returns score 0
entityExtractor() error
```

### CODE REVIEW

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 1]
```diff
@@ -32,22 +34,37 @@ case class EntityExtractorOutput(
...
 
   override def bodyParser(body: String, contentType: String, status: StatusCode): Map[String, String] = {
     if(StatusCodes.OK.equals(status)){
       val json = body.parseJson.asJsObject
-      val locationList = json.fields.get("LOC").map {
-        case JsArray(locations) => locations.toArray.map(_.toString.stripPrefix("\"").stripSuffix("\""))
-        case _ => Array.empty[String]
-      }.getOrElse(Array.empty[String])
+      // map over json, match entity_tye with case (non-null), create a list of strings
+      val extractedEntitiesMap = json.fields.filter(x => x._2.toString() != "null")
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:45]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

**Comment by angleto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466&bid=18376244)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala (hunk 1 above)
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:45]: Codacy Issue found: [Usage of untyped comparison operator](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 2]
```diff
@@ -32,22 +34,37 @@ case class EntityExtractorOutput(
...
+
+      val s = if(extractedEntitiesMap.isEmpty) "0" else "1"
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:47]: Codacy Issue found: [The string literal '0' appears 2 times in the file.](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 3]
```diff
@@ -32,22 +34,37 @@ case class EntityExtractorOutput(
...
+
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix("\"").stripSuffix("\"")))
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:50]: Codacy Issue found: [The string literal '"' appears 2 times in the file.](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 4]
```diff
@@ -32,22 +34,37 @@ case class EntityExtractorOutput(
...
+        case _ => List(List())
+      }.headOption.getOrElse(List())
 
-      val s = if(locationList.isEmpty) "0" else "1"
+      def outputMap(extractedEntities: List[Any], typeEntity: String) =
+        extractedEntities.zipWithIndex.map(x => (typeEntity + "." + x._2, x._1.toString)).toMap
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:55]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala (hunk 1 above)
Comment (bmichele) [reply src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:45]: @emanuelemaffeo How could I improve this one? Basically I need to check if the parameter `x._2: JsValue` is `null` but I did not manage to do it without converting it into a string.
Any suggestion?

**Comment by angleto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466&bid=18376244)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 5]
```diff
@@ -32,26 +34,47 @@ case class EntityExtractorOutput(
...
+      val s = if(extractedEntitiesMap.isEmpty) scoreNotExtracted else scoreExtracted
 
-      (locationList.indices.map(x => location + "." + x) zip locationList).toMap ++
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix("\"").stripSuffix("\"")))
+        case _ => List(List())
+      }.headOption.getOrElse(List())
+
+      def outputMap(extractedEntities: List[Any], typeEntity: String): Map[String, String] = {
+        val output = extractedEntities.zipWithIndex.map(x => (typeEntity + "." + x._2, x._1.toString))
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:59]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 6]
```diff
@@ -32,26 +34,47 @@ case class EntityExtractorOutput(
...
+      val jsonFields = body.parseJson.asJsObject.fields
+      val extractedEntitiesMap = jsonFields.filter(x => x._2.toString() =/= "null")
+
 
-      val s = if(locationList.isEmpty) "0" else "1"
+      val s = if(extractedEntitiesMap.isEmpty) scoreNotExtracted else scoreExtracted
 
-      (locationList.indices.map(x => location + "." + x) zip locationList).toMap ++
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix("\"").stripSuffix("\"")))
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:54]: Codacy Issue found: [The string literal '"' appears 2 times in the file.](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 7]
```diff
@@ -32,26 +34,47 @@ case class EntityExtractorOutput(
...
+    val scoreNotExtracted = "0"
+
     if(StatusCodes.OK.equals(status)){
-      val json = body.parseJson.asJsObject
-      val locationList = json.fields.get("LOC").map {
-        case JsArray(locations) => locations.toArray.map(_.toString.stripPrefix("\"").stripSuffix("\""))
-        case _ => Array.empty[String]
-      }.getOrElse(Array.empty[String])
+      val jsonFields = body.parseJson.asJsObject.fields
+      val extractedEntitiesMap = jsonFields.filter(x => x._2.toString() =/= "null")
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:48]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

**Comment by angleto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466&bid=18376244)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 8]
```diff
@@ -32,26 +34,54 @@ case class EntityExtractorOutput(
...
+      val quoteMark = "\""
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix(quoteMark).stripSuffix(quoteMark)))
+        case _ => List(List())
+      }.headOption.getOrElse(List())
+
+      def outputMap(extractedEntities: List[Any], typeEntity: String): Map[String, String] = {
+        val output = extractedEntities.zipWithIndex.map(x => {
+          val extractedEntityString = x._1.toString
+          (typeEntity + "." + x._2, extractedEntityString)
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:65]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 9]
```diff
@@ -32,26 +34,54 @@ case class EntityExtractorOutput(
...
-      (locationList.indices.map(x => location + "." + x) zip locationList).toMap ++
+      val quoteMark = "\""
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix(quoteMark).stripSuffix(quoteMark)))
+        case _ => List(List())
+      }.headOption.getOrElse(List())
+
+      def outputMap(extractedEntities: List[Any], typeEntity: String): Map[String, String] = {
+        val output = extractedEntities.zipWithIndex.map(x => {
+          val extractedEntityString = x._1.toString
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:64]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 10]
```diff
@@ -32,26 +34,54 @@ case class EntityExtractorOutput(
...
+
     if(StatusCodes.OK.equals(status)){
-      val json = body.parseJson.asJsObject
-      val locationList = json.fields.get("LOC").map {
-        case JsArray(locations) => locations.toArray.map(_.toString.stripPrefix("\"").stripSuffix("\""))
-        case _ => Array.empty[String]
-      }.getOrElse(Array.empty[String])
+      val jsonFields = body.parseJson.asJsObject.fields
+      val extractedEntitiesMap = jsonFields.filter(x => {
+        val jsonValueString = x._2.toString
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:49]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

**Comment by angleto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466&bid=18376244)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 11]
```diff
@@ -32,26 +34,55 @@ case class EntityExtractorOutput(
...
+
     if(StatusCodes.OK.equals(status)){
-      val json = body.parseJson.asJsObject
-      val locationList = json.fields.get("LOC").map {
-        case JsArray(locations) => locations.toArray.map(_.toString.stripPrefix("\"").stripSuffix("\""))
-        case _ => Array.empty[String]
-      }.getOrElse(Array.empty[String])
+      val jsonFields = body.parseJson.asJsObject.fields
+      val extractedEntitiesMap = jsonFields.filter(x => {
+        val jsonValue = x._2
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:49]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 12]
```diff
@@ -32,26 +34,55 @@ case class EntityExtractorOutput(
...
+      val quoteMark = "\""
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix(quoteMark).stripSuffix(quoteMark)))
+        case _ => List(List())
+      }.headOption.getOrElse(List())
+
+      def outputMap(extractedEntities: List[Any], typeEntity: String): Map[String, String] = {
+        val output = extractedEntities.zipWithIndex.map(x => {
+          val extractedEntity = x._1
+          val entityNumber = x._2
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:65]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

File: src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala [hunk 13]
```diff
@@ -32,26 +34,55 @@ case class EntityExtractorOutput(
...
-      (locationList.indices.map(x => location + "." + x) zip locationList).toMap ++
+      val quoteMark = "\""
+      val entityTypeAndList = extractedEntitiesMap.map {
+        case (entityType, JsArray(extractedEntities)) => (entityType, extractedEntities.toList.map(_.toString.stripPrefix(quoteMark).stripSuffix(quoteMark)))
+        case _ => List(List())
+      }.headOption.getOrElse(List())
+
+      def outputMap(extractedEntities: List[Any], typeEntity: String): Map[String, String] = {
+        val output = extractedEntities.zipWithIndex.map(x => {
+          val extractedEntity = x._1
```
Comment (angleto) [inline src/main/scala/com/getjenny/starchat/analyzer/atoms/http/custom/EntityExtractorVariableManager.scala:64]: Codacy Issue found: [Found unnamed result or parameter](https://app.codacy.com/gh/GetJenny/starchat/pullRequest?prid=5608466)

**Pull request merged**